File: src/cli/shell_completions.zig
pub const Shell = enum {
    unknown,
    bash,
    zsh,
    fish,
    pwsh,

    const bash_completions = @embedFile("completions-bash");
    const zsh_completions = @embedFile("completions-zsh");
    const fish_completions = @embedFile("completions-fish");

    pub fn completions(this: Shell) []const u8 {
        return switch (this) {
            .bash => bash_completions,
            .zsh => zsh_completions,
            .fish => fish_completions,
            else => "",
        };
    }

    pub fn fromEnv(comptime Type: type, SHELL: Type) Shell {
        // std.fs.path.basename only splits on '\' on Windows, so Unix-style
        // paths like "/usr/bin/bash" (set by Git Bash) return the full string.
        // Handle both separators and strip any ".exe" extension.
        const after_sep = if (std.mem.lastIndexOfAny(u8, SHELL, "/\\")) |i| SHELL[i + 1 ..] else SHELL;
        const basename = if (strings.hasSuffixComptime(after_sep, ".exe")) after_sep[0 .. after_sep.len - 4] else after_sep;
        if (strings.eqlComptime(basename, "bash")) {
            return .bash;
        } else if (strings.eqlComptime(basename, "zsh")) {
            return .zsh;
        } else if (strings.eqlComptime(basename, "fish")) {
            return .fish;
        } else if (strings.eqlComptime(basename, "pwsh") or
            strings.eqlComptime(basename, "powershell"))
        {
            return .pwsh;
        } else {
            return .unknown;
        }
    }
};

commands: []const []const u8 = &[_][]u8{},
descriptions: []const []const u8 = &[_][]u8{},
flags: []const []const u8 = &[_][]u8{},
shell: Shell = Shell.unknown,

pub fn print(this: @This()) void {
    defer Output.flush();
    var writer = Output.writer();

    if (this.commands.len == 0) return;
    const delimiter = if (this.shell == Shell.fish) " " else "\n";

    writer.writeAll(this.commands[0]) catch return;

    if (this.descriptions.len > 0) {
        writer.writeAll("\t") catch return;
        writer.writeAll(this.descriptions[0]) catch return;
    }

    if (this.commands.len > 1) {
        for (this.commands[1..], 0..) |cmd, i| {
            writer.writeAll(delimiter) catch return;

            writer.writeAll(cmd) catch return;
            if (this.descriptions.len > 0) {
                writer.writeAll("\t") catch return;
                writer.writeAll(this.descriptions[i]) catch return;
            }
        }
    }
}

const std = @import("std");

const bun = @import("bun");
const Output = bun.Output;
const strings = bun.strings;

File: src/cli/shell_completions.rs
use bun_core::Output;
use bun_paths;
use bun_str::strings;

#[repr(u8)]
#[derive(Copy, Clone, Eq, PartialEq, Default)]
pub enum Shell {
    #[default]
    Unknown,
    Bash,
    Zsh,
    Fish,
    Pwsh,
}

impl Shell {
    // TODO(port): @embedFile uses build-system module aliases ("completions-bash" etc.);
    // Phase B must wire the real relative paths or a build.rs OUT_DIR include.
    const BASH_COMPLETIONS: &'static [u8] = include_bytes!("completions-bash");
    const ZSH_COMPLETIONS: &'static [u8] = include_bytes!("completions-zsh");
    const FISH_COMPLETIONS: &'static [u8] = include_bytes!("completions-fish");

    pub fn completions(self) -> &'static [u8] {
        match self {
            Shell::Bash => Self::BASH_COMPLETIONS,
            Shell::Zsh => Self::ZSH_COMPLETIONS,
            Shell::Fish => Self::FISH_COMPLETIONS,
            _ => b"",
        }
    }

    // Zig: `fn fromEnv(comptime Type: type, SHELL: Type) Shell` — paired (comptime T, arg: T)
    // collapses to a single byte-slice param; callers pass `[]const u8`.
    pub fn from_env(shell: &[u8]) -> Shell {
        let basename = bun_paths::basename(shell);
        if basename == b"bash" {
            Shell::Bash
        } else if basename == b"zsh" {
            Shell::Zsh
        } else if basename == b"fish" {
            Shell::Fish
        } else if basename == b"pwsh" || basename == b"powershell" {
            Shell::Pwsh
        } else {
            Shell::Unknown
        }
    }
}

// File-level `@This()` struct.
// TODO(port): lifetime — fields are borrowed views (no deinit in Zig); Phase A forbids
// struct lifetime params, so using &'static. Phase B may need `<'a>`.
#[derive(Default)]
pub struct ShellCompletions {
    pub commands: &'static [&'static [u8]],
    pub descriptions: &'static [&'static [u8]],
    pub flags: &'static [&'static [u8]],
    pub shell: Shell,
}

impl ShellCompletions {
    pub fn print(&self) {
        let _flush = scopeguard::guard((), |_| Output::flush());
        let writer = Output::writer();

        if self.commands.is_empty() {
            return;
        }
        let delimiter: &[u8] = if self.shell == Shell::Fish { b" " } else { b"\n" };

        if writer.write_all(self.commands[0]).is_err() {
            return;
        }

        if !self.descriptions.is_empty() {
            if writer.write_all(b"\t").is_err() {
                return;
            }
            if writer.write_all(self.descriptions[0]).is_err() {
                return;
            }
        }

        if self.commands.len() > 1 {
            for (i, cmd) in self.commands[1..].iter().enumerate() {
                if writer.write_all(delimiter).is_err() {
                    return;
                }

                if writer.write_all(cmd).is_err() {
                    return;
                }
                if !self.descriptions.is_empty() {
                    if writer.write_all(b"\t").is_err() {
                        return;
                    }
                    if writer.write_all(self.descriptions[i]).is_err() {
                        return;
                    }
                }
            }
        }
    }
}

// ──────────────────────────────────────────────────────────────────────────
// PORT STATUS
//   source:     src/cli/shell_completions.zig (75 lines)
//   confidence: medium
//   todos:      2
//   notes:      @embedFile paths need build wiring; slice-of-slice fields may need <'a> in Phase B
// ──────────────────────────────────────────────────────────────────────────
